### PR TITLE
MetaMask support

### LINKED ISSUE #1131: [Demand] External Wallet Provider Interface
## Metadata

Field | Value
--- | ---
Previous Stage | ...
Figma Link | Figma_Link_Here____
Expected UX prototype due | 2020-W??-5
Actual UX prototype release | -
Expected engineering due | 2020-W??-5
Corresponding Pull Request | -
Completion Version | -

## Introduction

A user shall be able to connect Maskbook to external wallet providers.

When connected, the user shall be able to see wallets provided by the External Wallet Provider in the list of Wallets.

## General Principles

Local wallets will be marked as "Local". External wallets will be marked as "External". When no external wallet is added, both labels are invisible.

The External Wallet Provider can return an array of available wallets, where a wallet can already exist in the current Maskbook database as Local Wallet or External Wallet.

A wallet can come from multiple sources, including both Maskbook Local Wallet Database and External Wallet Providers. It is listed as multiple entries in the list of Wallets (Dashboard, wallet selection dialogue, etc).

## Notes

- Supporting MetaMask can be quick and easy, while others remain uncertain.

### PR DESCRIPTION
**Pull request by Jack-Works**
Leaving questions:

1. If user not installing MetaMask (https://github.com/MetaMask/extension-provider/issues/9)
2. Update wallet address when user change it in MetaMawsk (https://github.com/MetaMask/extension-provider/issues/11)
3. Declare we're Maskbook with a Logo in MetaMask (https://github.com/MetaMask/extension-provider/issues/12)
4. Get ERC20 token list that user added in the MetaMask (https://github.com/MetaMask/metamask-extension/issues/8865)
5. Get the name of the wallet from MetaMask (https://github.com/MetaMask/metamask-extension/issues/9336)

=====================

close #1131

image

image

image

### CODE REVIEW

**Comment by Jack-Works** [comment]
Gitcoin works!

File: src/extension/options-page/DashboardRouters/Wallets.tsx [hunk 1]
```diff
@@ -354,6 +350,12 @@ export default function DashboardWalletsRouter() {
             empty={!wallets?.length}
             title={t('my_wallets')}
             actions={[
+                <Button
+                    color="primary"
+                    variant="outlined"
+                    onClick={() => Services.Plugin.connectExoticWallet(WalletProviderType.metamask)}>
+                    Connect MetaMask
```
Comment (guanbinrui) [inline src/extension/options-page/DashboardRouters/Wallets.tsx:357]: i18n

File: src/plugins/Wallet/api.ts [hunk 2]
```diff
@@ -703,3 +698,6 @@ export const balanceCheckerAPI = (() => {
 type TxHashID = { transactionHash: string }
 type RedPacketID = { redPacketID: string }
 type DatabaseID = { databaseID: string }
+async function getRedPacketContract() {
```
Comment (guanbinrui) [inline src/plugins/Wallet/api.ts:701]: Plz do the same thing on other contracts.

File: src/plugins/Wallet/api.ts [hunk 3]
```diff
@@ -626,7 +619,7 @@ export const gitcoinAPI = {
                     onTransactionError: reject,
                     onEstimateError: reject,
                 },
-            )
+            ).catch(reject)
```
Comment (guanbinrui) [inline src/plugins/Wallet/api.ts:622]: The error should have caught in the `sendTx` function. No need to catch it again.

File: src/plugins/Wallet/tx.ts [hunk 4]
```diff
@@ -15,28 +15,28 @@ export async function sendTx<R, T extends TransactionObject<R>>(txObject: T, tx:
...
-                })
-                .on('receipt', (receipt: TransactionReceipt) => listeners?.onReceipt?.(receipt))
+        .then(([gas, gasPrice, nonce]) => {
+            const sent = txObject.send({
+                ...tx,
+                gas,
+                gasPrice,
+                nonce,
+            })
+            sent.catch((err: Error) => listeners.onTransactionError?.(err))
```
Comment (guanbinrui) [inline src/plugins/Wallet/tx.ts:25]: Chaining the `catch` function after `.on`?

File: src/plugins/shared/findOutProvider.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import type { WalletDetails } from '../../extension/background-script/PluginService'
+export enum WalletProviderType {
+    managed = 'managed',
```
Comment (guanbinrui) [inline src/plugins/shared/findOutProvider.ts:3]: What about renaming it to `maskbook`?

File: src/protocols/wallet-provider/index.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+export interface WalletProvider {
+    /** Check if the provider is available. */
+    checkAvailability(): Promise<boolean>
+    /**
+     * Get a list of wallets.
+     * This methods need user intention on external wallets!
+     */
+    requestAccounts(): Promise<string[]>
+    getWeb3Provider(): HttpProvider | IpcProvider | WebsocketProvider | AbstractProvider
+    noLongerUseWeb3Provider?(): void
```
Comment (guanbinrui) [inline src/protocols/wallet-provider/index.ts:19]: Comments are helpful!

File: src/plugins/shared/useWallet.ts [hunk 7]
```diff
@@ -55,10 +65,13 @@ export function useSelectWallet(wallets: WalletDetails[] | undefined, tokens: ER
     const erc20Balance = selectedToken
         ? `${formatBalance(selectedToken.amount, selectedToken.decimals)} ${selectedToken.symbol}`
         : undefined
+
+    const provider = selectedWallet ? findOutWalletProvider(selectedWallet) : undefined
     return {
         ethBalance,
         erc20Balance,
         //
```
Comment (guanbinrui) [inline src/plugins/shared/useWallet.ts:73]: Remove it.

File: src/plugins/Wallet/web3.ts [hunk 8]
```diff
@@ -1,60 +1,34 @@
...
-                  auto: true,
-                  delay: 5000, // ms
-                  maxAttempts: Number.MAX_SAFE_INTEGER,
-                  onTimeout: true,
-              },
-          })
-    if (pool.has(url)) provider.disconnect()
-    else pool.set(url, provider)
-    web3.setProvider(provider)
+let currentProvider: WalletProvider
```
Comment (guanbinrui) [inline src/plugins/Wallet/web3.ts:15]: You can delete this local variable read it from `lastActivatedWalletProvider.value`.

File: src/protocols/wallet-provider/maskbook.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            console.log('[Managed Web3] Switch', network)
+        },
+        disconnect() {
+            for (const each of pool.values()) each.disconnect()
+        },
+    }
+    extra.switch(currentLocalWalletEthereumNetworkSettings.value)
+    currentLocalWalletEthereumNetworkSettings.addListener((next) => extra.switch(next))
+    return new Proxy({} as HttpProvider & SwitchableProvider, {
+        get(_, p) {
```
Comment (guanbinrui) [inline src/protocols/wallet-provider/maskbook.ts:30]: Use more readable names.

File: src/protocols/wallet-provider/maskbook.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+import type { HttpProviderOptions } from 'web3-core-helpers'
+import type { EthereumNetwork } from '../../plugins/Wallet/database/types'
+import { getNetworkSettings } from '../../plugins/Wallet/UI/Developer/EthereumNetworkSettings'
+import { getManagedWallets } from '../../plugins/Wallet/wallet'
+import { web3 } from '../../plugins/Wallet/web3'
+import { currentLocalWalletEthereumNetworkSettings } from '../../settings/settings'
+import type { WalletProvider } from './index'
+
+type SwitchableProvider = { switch(network: EthereumNetwork): void; disconnect(): void }
+function SwitchableProvider(opts?: HttpProviderOptions): SwitchableProvider & HttpProvider {
```
Comment (guanbinrui) [inline src/protocols/wallet-provider/maskbook.ts:12]: I think you can achieve the goal with a local variable `pool`. Class and proxy make it harder to understand. Connecting and disconnecting can be easily done in `MaskbookProvider`.

File: src/plugins/Wallet/wallet.ts [hunk 11]
```diff
@@ -79,6 +84,56 @@ export async function getManagedWallets(): Promise<{
     return { wallets: await makeWallets(), tokens: makeTokens() }
 }
 
+export async function updateExoticWalletsFromSource(
```
Comment (guanbinrui) [inline src/plugins/Wallet/wallet.ts:87]: Plz divide the function into `updateExoticWalletFromSource` and `updateExoticWalletsFromSource`. The former one for the amniotic purpose.

File: src/protocols/wallet-provider/metamask.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+export const MetaMaskProvider: WalletProvider = {
+    async checkAvailability() {
+        try {
+            await timeout(MetamaskJSONRPC.eth_getBalance('0', 'latest'), 2000)
+            return true
+        } catch {
+            return false
+        }
+    },
+    async requestAccounts() {
```
Comment (guanbinrui) [inline src/protocols/wallet-provider/metamask.ts:46]: Do we really need all accounts? Or, only the one that the user is using now.

File: src/protocols/wallet-provider/metamask.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    {},
+    {
+        messageChannel: new EthereumJSONRpcChannel(MetaMaskWeb3Provider),
+        strict: false,
+        log: { remoteError: true },
+        key: 'Metamask',
+    },
+)
+
+export const MetaMaskProvider: WalletProvider = {
```
Comment (guanbinrui) [inline src/protocols/wallet-provider/metamask.ts:37]: Seems this provider doesn't support auto-syncing accounts.

File: src/extension/options-page/DashboardRouters/Wallets.tsx [hunk 14]
```diff
@@ -354,6 +350,12 @@ export default function DashboardWalletsRouter() {
             empty={!wallets?.length}
             title={t('my_wallets')}
             actions={[
+                <Button
```
Comment (guanbinrui) [inline src/extension/options-page/DashboardRouters/Wallets.tsx:353]: The button is better only visible when MetaMask was installed.

File: src/plugins/Wallet/wallet.ts (hunk 11 above)
Comment (Jack-Works) [reply src/plugins/Wallet/wallet.ts:87]: I don't understand

File: src/protocols/wallet-provider/maskbook.ts (hunk 10 above)
Comment (Jack-Works) [reply src/protocols/wallet-provider/maskbook.ts:12]: I'm not using class, what's the problem here?

File: src/plugins/Wallet/web3.ts (hunk 8 above)
Comment (Jack-Works) [reply src/plugins/Wallet/web3.ts:15]: `lastActivatedWalletProvider: InternalSettings<WalletProviderType>` is an enum, `currentProvider` is an instance, they are different. The prior one is used on the initial stage to preserve the choice of last running

File: src/plugins/shared/useWallet.ts (hunk 7 above)
Comment (Jack-Works) [reply src/plugins/shared/useWallet.ts:73]: This value is used

File: src/extension/options-page/DashboardRouters/Wallets.tsx (hunk 14 above)
Comment (Jack-Works) [reply src/extension/options-page/DashboardRouters/Wallets.tsx:353]: MetaMask/extension-provider#9

**Comment by Tedko** [comment]
Whats the progress so far besides https://github.com/MetaMask/extension-provider/issues/9 ? 
If ready to test, regardless of whether metaMask solved their issue we can have a try w/ users

File: src/extension/service.ts [hunk 15]
```diff
@@ -87,6 +88,13 @@ Object.assign(globalThis, {
     getCurrentNetworkWorkerService,
     ECKeyIdentifier,
     IdentifierMap,
+    BigNumber,
+})
+Object.defineProperty(BigNumber.prototype, '__debug__amount__', {
```
Comment (septs) [inline src/extension/service.ts:93]: the is global hacking. move to other file?

File: src/extension/service.ts [hunk 16]
```diff
@@ -102,7 +110,7 @@ function register<T extends Service>(service: T, name: keyof Services, mock?: Pa
                 log: logOptions,
                 channel: mc.eventBasedChannel,
                 preferLocalImplementation: GetContext() === 'background',
-                preservePauseOnException: process.env.NODE_ENV === 'development',
+                // preservePauseOnException: process.env.NODE_ENV === 'development',
```
Comment (septs) [inline src/extension/service.ts:113]: remove the line?

File: src/network/gun-servers.ts [hunk 17]
```diff
@@ -1,2 +1,2 @@
-export const gunServers = ['https://hk.gun.dimension.chat/gun', 'https://gun.maskbook.com/gun'] as const
+export const gunServers = ['https://gun.maskbook.com/gun'] as const
```
Comment (septs) [inline src/network/gun-servers.ts:1]: CC @yisiliu ?

File: src/extension/service.ts (hunk 16 above)
Comment (guanbinrui) [inline src/extension/service.ts:113]: Remove it.

File: src/plugins/Wallet/wallet.ts [hunk 18]
```diff
@@ -45,10 +46,16 @@ export async function isEmptyWallets() {
...
 
-export async function getWallets(): Promise<WalletRecord[]> {
-    const t = createTransaction(await createWalletDBAccess(), 'readonly')('Wallet', 'ERC20Token')
-    const wallets = await t.objectStore('Wallet').getAll()
-    return wallets.map(WalletRecordOutDB)
+export async function getWallets() {
+    const t = createTransaction(await createWalletDBAccess(), 'readonly')('Wallet')
+    const recs = await t.objectStore('Wallet').getAll()
+    const r = recs.map(WalletRecordOutDB).sort(sortWallet)
+    return r
```
Comment (guanbinrui) [inline src/plugins/Wallet/wallet.ts:53]:
```suggestion
    return recs.map(WalletRecordOutDB).sort(sortWallet)
```

**Review by septs — COMMENTED** [review]
Add a README file to `src/plugins/wallet`.
Reference `src/plugins/file-service/README.md`.

**Comment by Tedko** [comment]
@Jack-Works @guanbinrui 
for this question
```
Get ERC20 token list that user added in the MetaMask (MetaMask/metamask-extension#8865)

```

maybe we should just query from web3 provider like infura etc and dont need support from Metamask. eventually its the same thing, pull and display data from ETH blockchain

File: src/extension/service.ts (hunk 15 above)
Comment (Jack-Works) [reply src/extension/service.ts:93]: I think it's ok to do this

File: src/network/gun-servers.ts (hunk 17 above)
Comment (Jack-Works) [reply src/network/gun-servers.ts:1]: The first domain seems dead

**Review by guanbinrui — APPROVED** [review]

**Comment by Tedko** [comment]
About 3) 

From MetaMask team:


This is a good question. I believe during connection we have access to your extension ID, but it is not a name. We could do a lookup on the extension store for that browser and try to identify name & icon from there, but I believe we don't yet identify this information, so this is a good catch.

We're working overtime to launch 2 significant products this week, so it may be next week before we can dedicate our own resources to this issue, but I'll raise this with the team.

**Pull request merged**